src/hotspot/share/gc/g1/g1Policy.cpp

```diff
@@ -312,9 +307,14 @@ uint G1Policy::calculate_young_desired_length(size_t pending_cards, size_t rs_le
 // Limit the desired (wished) young length by current free regions. If the request
 // can be satisfied without using up reserve regions, do so, otherwise eat into
 // the reserve, giving away at most what the heap sizer allows.
-uint G1Policy::calculate_young_target_length(uint desired_young_length) const {
+uint G1Policy::calculate_young_target_length(uint desired_young_length, bool after_gc) const {
   uint allocated_young_length = _g1h->young_regions_count();
 
+  if (after_gc && (desired_young_length <= allocated_young_length)) {
+    log_trace(gc, ergo, heap)("Young target length: Desired young length smaller than allocated after GC. Allowing one extra eden region.");
+    desired_young_length = allocated_young_length + 1;
+  }
```

Comment:
The change forces the remaining code in `calculate_young_target_length` to add an extra region to the number of desired regions if the number of currently allocated regions is equal or exceeds them.
This override should only be used if the caller requested it, i.e. after a young gc to avoid an immediate full gc right away because the allocation that caused the gc could not be satisfied with zero eden regions.

E.g.
desired regions = 1 (because of very small heap size)
allocation regions = 1 (because of survivor regions from previous gc)

Now this leaves zero regions for eden for the next allocation, causing the described full gcs.
This code pretends to the actual target calculation (if after a gc) that we actually want one eden region.

As for the location of this `if`, I moved it to `calculate_young_desired_length` with a comment, since this code actually modifies the desired length. It's imho still a bit awkward there because it is kind of a tacked on condition, but I do not know a better place for this code (I've been fairly undecided about that in the original code alreayd).

Reply:
Thanks for the restructuring.  I think I understand what this is doing now.

I think the new structure is better, though I think the comment on the place
where the desired_young_length adjustment occurs could be improved. I think
what's happening there is that when computing the target after a GC, since we
want at least one eden region, and since allocated_young_length at that point
is the number of survivor regions, the desired length must exceed the
allocated (survivor) length.

But with that in mind, I think the prior name for the new argument
("after_gc") is better than the new name ("need_one_eden_region").  It's not
so much that the caller wants at least one eden region, it's that the policy
is that there should be at least one eden region after a GC.